fix(@desktop/wallet): Remove the 'Load More' button from the Activity view

Khushboo-dev-cpp opened the pull request:
… view and replace with automatic loading when the user scrolls down using a skeleton loading state

fixes #8987

raised a follow up for a glitchy behaviour seen https://github.com/status-im/status-desktop/issues/9320

### What does the PR do

Implements loading state + load more behavior for the transaction view as defined in design https://www.figma.com/file/FkFClTCYKf83RJWoifWgoX/Wallet-v2?node-id=10401%3A122250&t=k58Oy9eh3MsUsFer-4

When the transactions are loading we should see loading item and then on scrolling if the user drags the list, it should fetch more transaction
 
### Affected areas

Wallet

### StatusQ checklist

- [ ] add documentation if necessary (new component, new feature)
- [ ] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [ ] test changes in both light and dark theme?

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it



https://user-images.githubusercontent.com/60327365/214768818-edb499e6-4a74-4cd9-8f0b-872db68ecdb1.mov

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 65652fb6 | [#1](https://ci.status.im/job/status-desktop/job/branches/job/imports/job/PR-9098/1/) | 2023-01-12 23:33:11 | ~4 min | `imports` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/imports/job/PR-9098/1//consoleText) |

caybro commented on ui/imports/shared/controls/InformationTag.qml:
```diff
@@ -4,32 +4,44 @@ import QtQuick.Controls 2.14
+    horizontalPadding: Style.current.halfPadding
+    verticalPadding: 5
+
+    background: root.loading ? d.loadingComponent : root.customBackground
```
Why not make the `background` a simple `Loader` and choose its `sourceComponent` based on `root.loading`?

Khushboo-dev-cpp replied:
Can do too

caybro commented on ui/imports/shared/views/HistoryView.qml:
```diff
@@ -116,20 +94,53 @@ ColumnLayout {
             isIncoming: modelDataValid ? modelData.to === account.address: false
             cryptoValue: modelDataValid ? modelData.value : undefined
-            fiatValue: modelDataValid ? RootStore.getFiatValue(cryptoValue.amount, symbol, RootStore.currentCurrency) : undefined
+            fiatValue: modelDataValid  && cryptoValue !== undefined && !!cryptoValue ? RootStore.getFiatValue(cryptoValue.amount, symbol, RootStore.currentCurrency) : undefined
```
`cryptoValue !== undefined && !!cryptoValue`

Isn't that the same effectively?

Khushboo-dev-cpp replied:
Yep I just tried checking and it is effectively one and the same will remove

alaibe commented on src/app/modules/main/wallet_section/transactions/item.nim:
```diff
@@ -205,3 +223,6 @@ proc  getMaxTotalFees*(self: Item): CurrencyAmount =
   return self.symbol
+
+proc  getLoadingToken*(self: Item): bool =
+  return self.loadingToken
```
nit: loading transaction?

saledjenic commented on src/app/modules/main/wallet_section/transactions/model.nim:
```diff
@@ -213,8 +217,26 @@ QtObject:
+      if self.itemsWithDateHeaders[i].getLoadingToken():
+        self.beginRemoveRows(newQModelIndex(), i, self.itemsWithDateHeaders.len-1)
+        self.itemsWithDateHeaders.delete(i, self.itemsWithDateHeaders.len-1)
+        self.endRemoveRows()
```
@Khushboo-dev-cpp any reason why don't you emit `self.countChanged()` here?

saledjenic commented on src/app/modules/main/wallet_section/transactions/view.nim:
```diff
@@ -66,18 +67,24 @@ QtObject:
       self.models[address] = newModel()
 
     self.models[address].addNewTransactions(transactions, wasFetchMore)
+    if self.fetchingHistoryState[address] and wasFetchMore:
```
@Khushboo-dev-cpp there are few places here where we're accessing the value by key (for models and fetchingHistoryState) without checking if the key is present in the table, are we sure there is no case where in the moment of checking the searched key is not in the table?

Khushboo-dev-cpp replied:
I will double check this, I had tried to add wherever possible it will crash otherwise, thanks for the comment

Khushboo-dev-cpp commented:
@saledjenic @caybro I've addressed the comment, please check when you can :)

caybro commented on ui/imports/shared/controls/InformationTag.qml:
```diff
@@ -4,32 +4,46 @@ import QtQuick.Controls 2.14
+
+    QtObject {
+        id: d
+        property var loadingComponent: Component { LoadingComponent {}}
```
`readonly` as this never changes?

caybro reviewed (approved):
Looks good!

glitchminer commented:
Hi @Khushboo-dev-cpp , I was planning to test this by loading activity for a watch only account but was blocked by the add account functionality not being available. Please can you take a look? 
![image](https://user-images.githubusercontent.com/50769329/217104781-8b38a61d-f72e-4909-9245-caa67e20b0c0.png)

Build: https://ci.status.im/job/status-desktop/job/branches/job/macos/job/PR-9098/14/

Khushboo-dev-cpp commented:
> Hi @Khushboo-dev-cpp , I was planning to test this by loading activity for a watch only account but was blocked by the add account functionality not being available. Please can you take a look? ![image](https://user-images.githubusercontent.com/50769329/217104781-8b38a61d-f72e-4909-9245-caa67e20b0c0.png)
> 
> Build: https://ci.status.im/job/status-desktop/job/branches/job/macos/job/PR-9098/14/

Hey @glitchminer Ive rebased the PR again, I hoping it will work now :)

glitchminer commented:
### QA

Hi @Khushboo-dev-cpp , sorry for delay. Here are the notes from the test: 

---
1. OPEN "No transactions found" is shown even when there are transactions
![image](https://user-images.githubusercontent.com/50769329/217639819-b146ec1c-8f00-4957-920e-cfc2c23db3ea.png)

---
2. OPEN After a short delay where the loading images are shown 20 transactions are shown, on scroll no more transactions are loaded. Watched address: 0x56Eddb7aa87536c09CCc2793473599fD21A8b17F

https://user-images.githubusercontent.com/50769329/217641843-06583e80-a74d-4185-94d2-8ba7a4927bbd.mp4

Khushboo-dev-cpp commented:
> ### QA
> Hi @Khushboo-dev-cpp , sorry for delay. Here are the notes from the test:
> 
> 1. OPEN "No transactions found" is shown even when there are transactions
>    ![image](https://user-images.githubusercontent.com/50769329/217639819-b146ec1c-8f00-4957-920e-cfc2c23db3ea.png)
> 
> 2. OPEN After a short delay where the loading images are shown 20 transactions are shown, on scroll no more transactions are loaded. Watched address: 0x56Eddb7aa87536c09CCc2793473599fD21A8b17F
> 
>  2023-02-08_20-17-31.mp4

@glitchminer you need to drag and scroll the list up a little bit more. like you see in the video ive attached with the PR, if you think this not a normal behaviour, I can fine tune it further

glitchminer commented:
> @glitchminer you need to drag and scroll the list up a little bit more. like you see in the video ive attached here with this comment, if you think this not a normal behaviour, I can fine tune it further

Hi @Khushboo-dev-cpp , thanks, I've had it load more items now by clicking and dragging up. My initial understanding was that after just reaching the bottom then there would be a pause with the loading animation (recording in the previous comment is with a mouse scroll wheel).

I'm not sure it would be easy for a user to both identify and operate this function as is. Even once it's found, accessibility/dexterity issues may make it difficult for some to simultaneously click/drag. Have other methods the user might scroll been considered? For example,

    - mouse only users scrolling with scroll wheel
    - touchpad/mouse users scrolling by dragging the scroll bar
    - touchpad users with 2 finger scroll
    - keyboard users that want to use arrow keys or pg down

On another note, I saw comments in the issue stating that 25 items are expected to load but I only saw 20. Has the requirement changed?

Khushboo-dev-cpp commented:
> > @glitchminer you need to drag and scroll the list up a little bit more. like you see in the video ive attached here with this comment, if you think this not a normal behaviour, I can fine tune it further
> 
> Hi @Khushboo-dev-cpp , thanks, I've had it load more items now by clicking and dragging up. My initial understanding was that after just reaching the bottom then there would be a pause with the loading animation (recording in the previous comment is with a mouse scroll wheel).
> 
> I'm not sure it would be easy for a user to both identify and operate this function as is. Even once it's found, accessibility/dexterity issues may make it difficult for some to simultaneously click/drag. Have other methods the user might scroll been considered? For example,
> 
> ```
> - mouse only users scrolling with scroll wheel
> - touchpad/mouse users scrolling by dragging the scroll bar
> - touchpad users with 2 finger scroll
> - keyboard users that want to use arrow keys or pg down
> ```
> 
> On another note, I saw comments in the issue stating that 25 items are expected to load but I only saw 20. Has the requirement changed?

I think page size was never discussed 25/20.
Also about the loading behaviour let me ask @benjthayer  and @John-44 what they feel about this behaviour.
The users needs to hold and drag in order to loading more items. You can see that in the video I attached in the last comment. Should we change it to a simple scroll up till the end of list or current behaviour seems reasonable ?

glitchminer commented:
Hi @Khushboo-dev-cpp , just to confirm something before it's forgotten. Based on the feedback at the UI call on Friday, am I right in saying that the behaviour will be amended to load more on scrolling down only?

Khushboo-dev-cpp commented:
> ut the loading behaviour let me ask @benjthayer and @John-44 what they feel about this behaviour.
> The users needs to hold and drag in order to loading more items. You can see that in the video I attached in the last comment. Sh

Hey @glitchminer I have implemented the behaviour such that loading starts when we reach of list as discussed with Ben and John. Can you please retest this now?

glitchminer commented:
> > ut the loading behaviour let me ask @benjthayer and @John-44 what they feel about this behaviour.
> > The users needs to hold and drag in order to loading more items. You can see that in the video I attached in the last comment. Sh
> 
> Hey @glitchminer I have implemented the behaviour such that loading starts when we reach of list as discussed with Ben and John. Can you please retest this now?

Hi @Khushboo-dev-cpp , I've had another look at this now. The "No transactions found" issue has been resolved and the loading behaviour is as discussed, loading 20 more when the bottom has been reached.

Two issues were found, please confirm if the following are issues within the scope of this PR?
1. Once new activity has loaded the list jumps back to the top
2. Loading more activity does not take place when clicking and dragging the scrollbar

https://user-images.githubusercontent.com/50769329/220680923-78923514-f856-4c35-87d0-aa586b442138.mp4

https://user-images.githubusercontent.com/50769329/220680874-b392f2c0-d902-4bbd-87ca-888a2cfa7607.mp4

These issues are likely to be app wide and out of scope:
3. Scrolling using a fast scroll wheel.

Khushboo-dev-cpp commented:
> > > ut the loading behaviour let me ask @benjthayer and @John-44 what they feel about this behaviour.
> > > The users needs to hold and drag in order to loading more items. You can see that in the video I attached in the last comment. Sh
> > 
> > 
> > Hey @glitchminer I have implemented the behaviour such that loading starts when we reach of list as discussed with Ben and John. Can you please retest this now?
> 
> Hi @Khushboo-dev-cpp , I've had another look at this now. The "No transactions found" issue has been resolved and the loading behaviour is as discussed, loading 20 more when the bottom has been reached.
> 
> Two issues were found, please confirm if the following are issues within the scope of this PR?
> 
> 1. Once new activity has loaded the list jumps back to the top
> 2. Loading more activity does not take place when clicking and dragging the scrollbar
> 
> These issues are likely to be app wide and out of scope: 3. Scrolling using a fast scroll wheel. 4. Keyboard navigation using pg or arrow keys

@glitchminer thanks for the findings.
For 1. I have raised a new ticket to implement that as it is more complicated than the scope of this PR https://github.com/status-im/status-desktop/issues/9617
2. I have updated a fix for it.

test it when you can. Hoping all is good this time

glitchminer commented:
> 2. I have updated a fix for it.

@Khushboo-dev-cpp , it does load using the scrollbar now but it seems like it might be loading before getting to the bottom and/or getting stuck in a state where too many are loading at once.

https://user-images.githubusercontent.com/50769329/220768884-85e74350-b62e-44d0-8024-c6a6335adbc8.mp4

Khushboo-dev-cpp commented:
> > 2. I have updated a fix for it.
> 
> @Khushboo-dev-cpp , it does load using the scrollbar now but it seems like it might be loading before getting to the bottom and/or getting stuck in a state where too many are loading at once.

Hey @glitchminer tried another fix that is closer to start loading at end of lists scroll. Please note I see a jumping behaviour when using scrollbar and thats just cause the scrollbar height gets recalculated cant really be helped.

The getting stuck is not something I noticed, it maybe cause the UI was blocked due to some other background process as fetching history items is a async api call.

glitchminer commented:
Thanks @Khushboo-dev-cpp , the freeze where it never actually loads the list only seems to happen when it's the first run of the instance. After a restart it resolves. This is the current behaviour when dragging the scrollbar. Is the fact that it jumps to the middle of the next batch of transactions an issue?

https://user-images.githubusercontent.com/50769329/220911264-1b1ec3c2-fb82-40e2-8190-5456fce4bd10.mp4

Khushboo-dev-cpp commented:
> Thanks @Khushboo-dev-cpp , the freeze where it never actually loads the list only seems to happen when it's the first run of the instance. After a restart it resolves. This is the current behaviour when dragging the scrollbar. Is the fact that it jumps to the middle of the next batch of transactions an issue?

The issue is we are scrolling as we keep growing the list and thats whats causing this jumping behaviour as the scrollbar has to resize. Not sure if there could be a fix for this as we are using Qt's default scrollbar. But we can raise an issue and see if something can be done

glitchminer commented:
OK, thanks for the clarification @Khushboo-dev-cpp. We can deal with it as a separate issue as needed.

glitchminer reviewed (approved):
Tested, approved
